== four categories of errors:

* standard javascript errors
  - EvalError
	- SyntaxError
	- RangeError
	- ReferenceError
	- TypeError
	- URIError
* system errors
* user-specified errors
* assertion error

<Error>, the base class
	
== error propagation and interception

=== mechanisms for propagating and handling errors

* the type of Error
* the style of the API

=== generate and throw an error in sync/async code

```
  try {
		throw new Error();
  } catch (err) {
		console.log(err);
  }
```

=== the first argument as an Error object in callbacks

```
  const fs = require('fs');
	fs.readFile('a file that does not exist', (err, data) => {
		if (err) {
			console.error('There was an error reading the file!', err);
			return;
		}
	});
```

=== errors are routed to the 'error' event in an EventEmitter

```
  const net = require('net');
	
	const connection = net.connect('localhost');
	connect.on('error', (err) => {
		console.error(err);
	});
	
	connection.pipe(process.stdout);
```

=== register a handler for the process.on('uncaughtException')

== Node.js style callbacks

the try/catch cannot be used to intercept errors generated by asynchronous APIs

```
  const fs = require('fs');
	
	try {
		fs.readFile('/some/file/that/does-not-exist', (err, data) => {
			// mistaken assumption: throwing here ...
			if (err) {
			  throw err;
			}
		});
	} catch (err) {
		// This will not catch the throw!
		console.error(err);
	}
```
